tracee-ebpf/tracee/events_derived.go

```diff
@@ -0,0 +1,92 @@
+package tracee
+
+import (
+	"fmt"
+
+	"github.com/aquasecurity/tracee/pkg/external"
+)
+
+type deriveFn func(EventDefinition) error
+
+func (t *Tracee) deriveEvent(event external.Event) []external.Event {
+	var deriveFns map[int32]deriveFn
+	var derivatives []external.Event
+
+	eventID := int32(event.EventID)
+	switch eventID {
+	case CgroupMkdirEventID:
+		containerCreateFn := func(def EventDefinition) error {
+			cgroupId, err := getEventArgUint64Val(&event, "cgroup_id")
+			if err != nil {
+				return err
+			}
+
+			if info := t.containers.GetCgroupInfo(cgroupId); info.ContainerId != "" {
+				de := event
+				de.EventID = int(ContainerCreateEventID)
+				de.EventName = "container_create"
+				de.ReturnValue = 0
+				de.StackAddresses = make([]uint64, 1)
+				de.Args = []external.Argument{
+					{ArgMeta: def.Params[0], Value: info.Runtime},
+					{ArgMeta: def.Params[1], Value: info.ContainerId},
+					{ArgMeta: def.Params[2], Value: info.Ctime},
+				}
+				de.ArgsNum = 3
+
+				derivatives = append(derivatives, de)
+			}
+
+			return nil
+		}
+		deriveFns = map[int32]deriveFn{
+			ContainerCreateEventID: containerCreateFn,
+		}
+	case CgroupRmdirEventID:
+		containerRemoveFn := func(def EventDefinition) error {
```

Comment:
i think these inner funcs could be defined in the outer scope, so the `deriveEvent()` would be smaller - as i think we'll be adding more higher level events in the future. but it's your call :)

Reply:
I thought about it too, but having it this way, we're saving passing the original event as an argument. In addition, the interface for `deriveFn` would have to be changed to return the derived event, which will have to add a check that the event is not empty (in the case that there is no error, but also no event has been derived, e.g. when `containerID == ""`)

Reply:
Yep, I thought the same and then saw that Yaniv was using event from the outer function.